src/builder/QuarkBuilder.sol

```diff
@@ -989,6 +1000,48 @@ contract QuarkBuilder {
         return balanceOnChain < amountNeededOnDstChain;
     }
 
+    /**
+     * @dev Check if the asset required its wrapped/unwrapped counterpart balance to cover the intented action
+     * and if so, insert wrap/unwrap action to cover the original intent amount
+     */
+    function checkAndInsertWrapOrUnwrapAction(
+        List.DynamicArray memory actions,
+        List.DynamicArray memory quarkOperations,
+        Accounts.ChainAccounts[] memory chainAccountsList,
+        PaymentInfo.Payment memory payment,
+        string memory assetSymbol,
+        uint256 amount,
+        uint256 chainId,
+        address account,
+        uint256 blockTimestamp,
+        bool useQuotecall
+    ) internal pure {
+        // Check if inserting wrapOrUnwrap action is necessary
+        uint256 balanceOnOriginalAsset = Accounts.getBalanceOnChain(assetSymbol, chainId, chainAccountsList);
+        if (balanceOnOriginalAsset < amount && TokenWrapper.hasWrapperContract(chainId, assetSymbol)) {
+            // If the asset has a wrapper counterpart, wrap/unwrap the token to cover the transferIntent amount
+            string memory counterpartSymbol = TokenWrapper.getWrapperCounterpartSymbol(chainId, assetSymbol);
+
+            // Wrap/unwrap the token to cover the transferIntent amount
+            (IQuarkWallet.QuarkOperation memory wrapOrUnwrapOperation, Actions.Action memory wrapOrUnwrapAction) =
+            Actions.wrapOrUnwrapAsset(
+                Actions.WrapOrUnwrapAsset({
```

Comment:
This should always work because `assertFundsAvailable` already checks that there are enough of the counterpart asset on the destination chain right?

Reply:
Yeah. 
When the builder reaches this step, whatever that's on the dstChain should have enough to cover what we needs.  Otherwise it should fail in `assertFundsAvailable` as it should check on available funds across chains and including the wrapped/unwrapped version of the token on dst chain.